### PR TITLE
Tag Plots and PlotUtils

### PR DESCRIPTION
**Pull request by tbreloff**
I updated PlotUtils require to julia 0.5-.  For Plots changes, see the [NEWS](https://github.com/tbreloff/Plots.jl/blob/master/NEWS.md)

### CODE REVIEW

**Comment by tkelman** [comment]
You really shouldn't be doing `Pkg.add` in package code.

**Comment by tbreloff** [comment]
I'll take that out once we have a package manager and conditional dependencies in Base Julia that can ensure easy installs for users.  But I have absolutely no hesitation putting appropriate `Pkg.add` calls inside a method named `add_backend`.

**Comment by tkelman** [comment]
That's not a package's job, that's the package manager's job. Why is `Plots.add_backend("PyPlot")` superior to `Pkg.add("PyPlot")` ?

**Comment by tbreloff** [comment]
`add_backend` is intended to be the "I'm lost, just set it up for me" approach for new users (and for me, when I don't feel like remembering the "right way" to install a package every time).  For pyplot, this forces the standard Conda.jl install by setting `ENV["Python"] = ""` before building.  For plotlyjs, it adds the optional, but highly recommended Rsvg package as well as doing the `Blink.AtomShell.install()` that has prompted many new users to open issues.  For glvisualize... well lets just say I wouldn't be able to run @SimonDanisch's code without that method ;)

tl;dr The package manager doesn't make it easy enough, so I'm supplementing

**Comment by tkelman** [comment]
If it's only ever intended to be called manually, then it really shouldn't be part of the package's API. Now there's a function that every future tag of Plots.jl should be checked that there's no unconditional calling of it, mutating the user's package state. Likewise, all other packages that depend on Plots.jl should be checked that they're not calling it in package code.

How about implementing it as a non-default code path in `deps/build.jl`, so you do something like set `ENV["PLOTSJL_BACKEND"] = "PyPlot"; Pkg.build("Plots")` ?

**Comment by tbreloff** [comment]
> How about implementing it as a non-default code path in deps/build.jl, so you do something like set ENV["PLOTSJL_BACKEND"] = "PyPlot"; Pkg.build("Plots") ?

Setting environment variables and manually rebuilding?  That's about as bad as the annoyances I'm trying to alleviate.   If I don't export the method, does that make you more comfortable?  Or would you prefer a different name, like `easy_install`?  

Do you want me to print a warning and ask for confirmation?

```
WARNING: This will install the requested backend and may add new packages and/or checkout development branches of existing packages... do you still wish to proceed?  y/n
```

There are common stumbling blocks for getting new users installed and ready... we as a community should do everything we can to make that first experience seamless.

**Comment by tkelman** [comment]
It definitely shouldn't be exported. I don't care about the name, I care about whether it's in the module such that it would ever get called without the user directly asking for it. Asking for confirmation is a pretty good idea. Please don't check out master branches of anything, *especially* if these are supposed to be for new users. If something doesn't work on a tagged release, it's not ready for users.

If you don't like making it called via Pkg.build, then you could put them in separate tiny jl files where you don't include them in the main Plots module but do set a variable name for the jl file name: set `Plots.add_PyPlot = joinpath(@__DIR__, "add_PyPlot.jl")` then the user does `include(Plots.add_PyPlot)`

**Comment by tbreloff** [comment]
Ok you win.  I redesigned add_backend to just output a helpful copy/paste string.  And I also added the call the add_backend inside backend initialization.  I uninstalled PlotlyJS, and this is the flow for a new user now:

```julia
julia> using Plots; plotlyjs(); plot(rand(10))
INFO: Recompiling stale cache file /home/tom/.julia/lib/v0.5/Plots.ji for module Plots.
WARNING: Couldn't initialize plotlyjs.  (might need to install it?)
INFO: To do a standard install of plotlyjs, copy and run this:

if !Plots.is_installed("PlotlyJS")
    Pkg.add("PlotlyJS")
end
if !Plots.is_installed("Rsvg")
    Pkg.add("Rsvg")
end
import Blink
Blink.AtomShell.install()


ERROR: ArgumentError: Module PlotlyJS not found in current path.
Run `Pkg.add("PlotlyJS")` to install the PlotlyJS package.
 in require(::Symbol) at ./loading.jl:365
 in _initialize_backend(::Plots.PlotlyJSBackend) at /home/tom/.julia/v0.5/Plots/src/backends/plotlyjs.jl:28
 in backend() at /home/tom/.julia/v0.5/Plots/src/backends.jl:151
 in #plotlyjs#172(::Array{Any,1}, ::Function) at /home/tom/.julia/v0.5/Plots/src/backends.jl:20
 in plotlyjs() at /home/tom/.julia/v0.5/Plots/src/backends.jl:20

julia> if !Plots.is_installed("PlotlyJS")
           Pkg.add("PlotlyJS")
       end
INFO: Installing DocStringExtensions v0.2.1
INFO: Installing PlotlyJS v0.4.2
INFO: Building HttpParser
INFO: Building MbedTLS
INFO: Package database updated
INFO: METADATA is out-of-date — you may not have the latest version of PlotlyJS
INFO: Use `Pkg.update()` to get the latest versions of your packages

julia> if !Plots.is_installed("Rsvg")
           Pkg.add("Rsvg")
       end
INFO: Installing Rsvg v0.0.2
INFO: Building Cairo
INFO: Building Gtk
WARNING: `@linux_only` is deprecated, use `@static if is_linux()` instead
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in @linux_only(::Any) at ./deprecated.jl:500
 in include_from_node1(::String) at ./loading.jl:488
 in evalfile(::String, ::Array{String,1}) at ./loading.jl:504 (repeats 2 times)
 in cd(::##2#4, ::String) at ./file.jl:59
 in (::##1#3)(::IOStream) at ./none:13
 in open(::##1#3, ::String, ::String) at ./iostream.jl:113
 in eval(::Module, ::Any) at ./boot.jl:234
 in process_options(::Base.JLOptions) at ./client.jl:239
 in _start() at ./client.jl:318
while loading /home/tom/.julia/v0.5/Gtk/deps/build.jl, in expression starting on line 16
INFO: Building Rsvg
INFO: Package database updated
INFO: METADATA is out-of-date — you may not have the latest version of Rsvg
INFO: Use `Pkg.update()` to get the latest versions of your packages

julia> import Blink
WARNING: Method definition require(Symbol) in module Base at loading.jl:345 overwritten in module Main at /home/tom/.julia/v0.5/Requires/src/require.jl:12.

julia> Blink.AtomShell.install()
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  386k  100  386k    0     0   713k      0 --:--:-- --:--:-- --:--:--  712k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   159    0   159    0     0    534      0 --:--:-- --:--:-- --:--:--   535
100   606    0   606    0     0   1694      0 --:--:-- --:--:-- --:--:--  1694
100 35.2M  100 35.2M    0     0  6996k      0  0:00:05  0:00:05 --:--:-- 8376k

julia> plot(rand(10))
```

**Comment by tbreloff** [comment]
For completeness:

```julia
julia> for b in Plots.backends()
           add_backend(b)
       end
INFO: To do a standard install of pyplot, copy and run this:

if !Plots.is_installed("PyPlot")
    Pkg.add("PyPlot")
end
withenv("PYTHON" => "") do
    Pkg.build("PyPlot")
end


INFO: To do a standard install of unicodeplots, copy and run this:

Pkg.add("UnicodePlots")
Pkg.build("UnicodePlots")


INFO: To do a standard install of plotly, copy and run this:

Pkg.build("Plots")


INFO: To do a standard install of plotlyjs, copy and run this:

if !Plots.is_installed("PlotlyJS")
    Pkg.add("PlotlyJS")
end
if !Plots.is_installed("Rsvg")
    Pkg.add("Rsvg")
end
import Blink
Blink.AtomShell.install()


INFO: To do a standard install of gr, copy and run this:

Pkg.add("GR")
Pkg.build("GR")


INFO: To do a standard install of glvisualize, copy and run this:

if !Plots.is_installed("GLVisualize")
    Pkg.add("GLVisualize")
end
if !Plots.is_installed("Contour")
    Pkg.add("Contour")
end
if !Plots.is_installed("GLPlot")
    Pkg.add("GLPlot")
end

# TODO: remove this section when the tagged versions catch up
for pkg in [
        "GLWindow", "GLAbstraction",
        "GLVisualize", "GeometryTypes", "FixedSizeArrays",
        "FreeType", "GLPlot"
    ]
    warn("Running Pkg.checkout("$pkg").  To revert, run Pkg.free("$pkg")")
    Pkg.checkout(pkg)
end
warn("Running Pkg.checkout("Reactive", "sd/betterstop").  To revert, run Pkg.free("Reactive")")
Pkg.checkout("Reactive", "sd/betterstop")


INFO: To do a standard install of pgfplots, copy and run this:

Pkg.add("PGFPlots")
Pkg.build("PGFPlots")
```

**Comment by tbreloff** [comment]
I just pushed up 0.9.3 with the add_backend redesign... people are waiting on this tag to get merged, so hopefully there are no more blockages.

**Comment by tkelman** [comment]
I don't think you should be recommending that people check out master or random branches of that many packages. If the GLVisualize stack needs tags, then tag them.

**Comment by tbreloff** [comment]
Thank you for merging @tkelman.  I will only recommend checking out master or random branches when it's necessary.

**Comment by tkelman** [comment]
My point is you shouldn't recommending it at all. Get the work done to make it no longer necessary, until that work is done it's not ready for users.
